### DISCUSSION TITLE
How to mark imports "external" in loader?

### DISCUSSION THREAD

**Discussion by sagargurtu** [webpack/webpack#18184] [Q&A]
I'm creating a webpack loader that transforms source code and adds/modifies import statements. For example:
```
// The following import is injected in `content`
// Library type is 'module'

import "/foo/bar.js";
```

This import needs to be left untouched and marked "external". These imports are only known at runtime and cannot be added under `externals` option in webpack config. How can this be achieved from the loader function?

**Comment by alexander-akait** [comment]
You can't do it in a loader

**Reply by sagargurtu** [reply]
Any suggestions on how to do both (code transformation and marking imports external) via plugin hooks if possible?

**Reply by alexander-akait** [reply]
Yes, you can modify `externals` options in your plugin and put required things there, then add your loader and handle it

**Reply by sagargurtu** [reply]
Thanks @alexander-akait 

- The import statements will only be known during code transformation. Will modifying `externals` option still work at that time?
- `externals` option may already be populated by the user. Is there any way the plugin can programmatically mark new imports external? Which hook(s) should I use for this?
- Which plugin hook(s) should I use for source code transformation? Or is there a way to inject the loader from the plugin? (I can have only one of plugin or loader)

In short, I'm looking for a way to transform the source code, inject new import statements (that will only be known during code transform), and mark them external, all from within a plugin.

**Reply by alexander-akait** [reply]
> The import statements will only be known during code transformation. Will modifying externals option still work at that time?
> externals option may already be populated by the user. Is there any way the plugin can programmatically mark new imports external? Which hook(s) should I use for this?

Based on what you want, but yes, it will work, also you can make a communication between plugin and loader (just pass an option or store it in buildInfo or you can use symbol for loader content and store values where), then you modify externals options and webpack will ignore them

There is an example plugin + loader https://github.com/webpack-contrib/image-minimizer-webpack-plugin/blob/master/src/index.js#L465, you can get ideas from there, feel free to feedback
